### ISSUE TITLE
Dissonance/Integrations folder missing on Asset Store

### ISSUE THREAD

**Issue by endovert** [Placeholder-Software/Dissonance#39]
## Context
Downloaded the asset and trying to follow the tutorial:
http://dissonance.readthedocs.io/en/latest/Basics/Quick-Start-UNet-HLAPI/

## Expected Behavior
There should be Dissonance/Integrations folder in the downloaded assets, 
 and other sample scene files.

## Actual Behavior
There is no Dissonance/Integrations folder in the imported asset.
No unity scene file is present in the downloaded asset.

## Steps to Reproduce
1. Download the asset from Unity Asset Store
2. Look for the Dissonance/Integrations folder
3. Not there

## Your Environment
- **Dissonance version used**: v1.1.1

- **Unity version**: 5.5.1f1

- **Editor Operating System and version**: Microsoft Window 10

- **Build Settings**: N/A

- **Link to your project**: N/A

**Comment by martindevans** [comment]
Hi endovert. It looks like you've missed a step from your setup, check the [Project Setup](https://dissonance.readthedocs.io/en/latest/Basics/Getting-Started/) page on the docs. Particularly this bit:

> Import the Dissonance asset into the project. Once Dissonance has installed and the project has compiled a window will pop up offering to take you to the integrations list - visit this page to see the latest changelog and to download the optional integrations for Dissonance

The download page for the current version is [here](https://placeholder-software.co.uk/dissonance/releases/1.1.1.html).

Did you not get that popup window when you first installed Dissonance? If not this is a bug which we definitely need to fix!

**Comment by endovert** [comment]
Oh, I see. My bad. 

Suggestion: it may be clearer if you break down the first step in the Project Setup page into two (1. import asset + 2. Download integrations) for those who are not paying attention like me.

Thanks!

**Comment by martindevans** [comment]
Good idea. I've rewritten the docs to include a specific `Download Integration` subsection. You can see it [here](https://github.com/Placeholder-Software/Dissonance/blob/master/docs/Basics/Getting-Started.md). Unfortunately it's not live yet, out docs host seems to be playing up.

**Comment by morderkaine1** [comment]
I have tried to update to the latest version and had a similar problem - the window does not pop up and also there is no Dissonance option under the Window tab in Unity so I cannot follow the instructions at https://dissonance.readthedocs.io/en/latest/Basics/Getting-Started/ as I cannot get the Integrations folder.

**Comment by martindevans** [comment]
That's odd - does the project you imported it into fail to build at the moment? The window is part of the project, so if it isn't successfully building then it won't show the welcome window. If it's not building, is the error in Dissonance? And if so, what's the error and what editor version are you using?

The release page for each version of Dissonance can be found [here](https://placeholder-software.co.uk/dissonance/releases/). Make sure to use the correct version - the integrations may not be compatible with other versions!

**Comment by morderkaine1** [comment]
I did have one error in an extra script that I thought I deleted before
importing, maybe I did it after. Perhaps that prevented the welcome page
from opening - or I tried going from way too old of a version to the new
one.
Once I got the integrations added by saving the link to the download
location everything did work fine. I have also been having some issues with
MonoDevelop complaining about errors that don't really exist (at one point
I had to move the HLAPIServer and Client scripts out of their folder and
back in and that miraculously fixed some errors MonoDevelop was complaining
about).

My recommendation in case other people have a similar issue is to have a
link to the Integrations page in the Setup instructions. As soon as I
imported that bit, everything worked right and Dissonance appeared as an
option in the Window tab in Unity.

**Comment by martindevans** [comment]
Glad to hear you got it working :)

> My recommendation in case other people have a similar issue is to have a link to the Integrations page in the Setup instructions

I'm investigating if I can move the code associated with the welcome system into a precompiled DLL, so it will show up even if the project is not compiling. I'll fall back to adding a link into the docs if I can't get the standalone welcome window to work.

Having it in the docs isn't ideal for a number of reasons. People don't always read the docs. People _upgrading_ Dissonance definitely won't read the setup docs but they do need a reminder about downloading new integrations. Also the version is critical (the welcome window takes you to the right version) and I'd rather not rely on people taking themselves to the right version!

**Comment by morderkaine1** [comment]
Good points!

**Comment by morderkaine1** [comment]
Hi Martin,
I ran into trouble again. For some reason after running the game after
trying the update to the latest Dissonance after updating Unity, Dissonance
created an extra plugins folder. So there was a file structure like Assets
-> Plugins -> Dissonance -> Plugins
And then when I try to build it complains about two DLL files interfering
with each other. Pretty sure that second Plugins folder should not be
there, but it gets automatically created somehow.

What I am going to try and do this evening is remove everything to do with
Dissonance and reinstall from nothing. That will mean the two Dissonance
folders, the Dissonnance game object with the broadcast and receipt
scripts, and is there anything else I would have to do to clear it out
completely?

**Comment by martindevans** [comment]
The second Plugins folder is definitely meant to be there! The two locations have slightly different meanings:
 - `Assets/Plugins` is a special location which Unity uses to mean that this code is not part of the game, it is precompiled before anything else and is not recompiled when other code changes. This means that when you change a script which is not in plugins the compile happens much faster (because it doesn't need to recompile everything in `Assets/Plugins`). So naturally we put the bulk of the Dissonance code into here to speed up compilation for all our users.
 - `Assets/Plugins/Dissonance/Plugins` contains the [native plugins](https://docs.unity3d.com/Manual/NativePlugins.html) which Dissonance uses (opus codec, audio preprocessing etc).

I would suggest deleting and re-importing Dissonance. You shouldn't have to delete any scripts in your scenes (the gameobjects *should* pick up the scripts again after you re-import). If that doesn't fix it:
 - Which two plugins does it say are conflicting?
 - Which platform are you building for?
 - Which version of Dissonance?
 - Which version of Unity?

**Comment by morderkaine1** [comment]
Hi,
it says Plugin 'AudioPluginDissonance.dll' is used from several locations:
Building for Windows, UNET HLAPI
The version of Dissonance on the asset store as the most current - #3.0.2
And Unity 5.6.3f1

Looking at what happened in my last install attempt here are the two folder
structures:
Assets -> Dissonance -> Plugins
Assets -> Plugins -> Dissonance

I'm guessing the first plugins folder should be moved so its Assets ->
Plugins -> Dissonance -> Plugins as per your last reply?

Found more details also playing with it - there was a
AudioPluginDissonance.dll in these two places
Assets -> Plugins -> x86
 Assets -> Plugins -> Dissonance -> Plugins -> x86 (and also x86_64)
Removing it from the first one lets me build and prevents the conflicting
plugins error

**Comment by martindevans** [comment]
Aha that's definitely the problem, that `Assets -> Dissonance -> Plugins` shouldn't be there. You should have

`Assets/Dissonance/Core`
`Assets/Dissonance/Integrations/`

And then in Plugins you should have a number of folders (Core, resources etc). Check that `Assets/Plugins/Dissonance/Plugins` is there and contains subfolders for different architectures.

**Comment by morderkaine1** [comment]
So I have it working now for global chat (which when I am testing team chat
is set to PushToTalk so it doesnt interfere) however when I set up team
chat its broadcasting to everyone.
I'm using a player script to change the room in the Broadcast and Receive
scripts. In the scripts it is being set properly however when I run two
instances of the game locally even though the two players are on different
teams I can hear myself talking.
For example, I have a red and blue team and have code like
teamVoiceChatReceiver.RoomName = "Red Team";
teamVoiceChatBroadcast.RoomName = "Red Team";
and when on the Blue team same thing but "Blue Team".

**Comment by morderkaine1** [comment]
Do you know why when one player is set to Red Team room and the other
player Blue Team they can hear each other talk? It's the one part that I
cant get working correctly right now.

**Comment by martindevans** [comment]
Hi morderkaine, sorry for not replying to this one - I typed out a reply this morning and it looks like I never pressed submit!

I think the problem is that the receiver only applies the new room when it *starts* receiving. So you set the `RoomName` property from "Global" to "Red Team" on the receiver, but that change never gets applied because the receiver was already listening and never stops listening! You can verify this by increasing the log level (`Window > Dissonance > Diagnostic Settings`) on "Core" to "Debug". When you start and stop listening to a room you will see messages like this:

```
DEBUG [Dissonance:Core] Rooms: Joined chat room 'Room Name'
DEBUG [Dissonance:Core] Rooms: Left chat room 'Room Name'
```

So if I'm right you will currently *not* see these messages when your script runs. If that's the case then you can change the `RoomName` property of `VoiceReceiptTrigger` to:

```
public string RoomName
{
    get { return _roomName; }
    set
    {
        if (_roomName != value)
        {
        _roomName = value;

        if (_membership.HasValue)
        {
            Comms.Rooms.Leave(_membership.Value);
            _membership = null;
        }
    }
}
```

With that in place you should see the message pop up in the log when your script runs, and you should not have channels working correctly. I've already submitted a patch to fix this and it should be in the next release of Dissonance.

On a side note - why even change the room name? You could just have two receipt triggers (i.e. team and global) and enable/disable them as needed. Alternatively you could manually join the room like so:

```
DissonanceComms comms;

var membership = comms.Rooms.Join("Red Team");
// Now you're listening to the red team room

comms.Rooms.Leave(membership);
//And now you're not
```

**Comment by morderkaine1** [comment]
Hi,
Thanks for the tips, I will try them!
The implementation I had with a (much) earlier version worked with this
method but I guess things have changed. If they never stop listening (you
cant just switch rooms at any moment and have it be instantly applied) then
my implementation wont work.

I change the room name because players will switch teams in between rounds
during games. So best option is to have 3 triggers, global as a push to
chat, red and blue as voice activation and turn them on and off as the
player switches teams?
It seems a bit easier to understand the process than the manual joining
rooms code.

**Comment by morderkaine1** [comment]
So I did a more involved test, and I think there is a bug somewhere.

The only modifications to your code I made was in  private bool
IsUserActivated()

case CommActivationMode.PushToTalk:
return pushToTalk;

where pushToTalk is a bool I set via a button press, but only the Global
chat is set to push to talk. Global chat scripts were turned off for this
test.

I set up 3 receipt and 3 broadcast scripts, red, blue and global. Global
turned off, and scripts to turn on/off the team ones.

I loaded up a build and then loaded in the editor and connected to the
build.
I saw in the editor that only one teams receipt and broadcast trigger
scripts were active, the other teams was turned off. my player code
randomly selects between blue and red and keeps the teams even.

I did tests of turning off all receipt/broadcast scripts then turning them
on one by one. I know from viewing the start the editor that the other
player had only 1 teams broadcast and receipt on. As long as I had 1 single
script of either team, of either broadcast or receiver I could hear myself
talk. Only when one player had ALL team scripts turned off could I not hear
voice.
Basically as far as I can tell, there is no difference for voice activation
for being in different rooms, all rooms hear each other.
